storagedriver/azure/azure.go

```diff
@@ -174,11 +193,17 @@ func (d *Driver) WriteStream(path string, offset, size int64, reader io.ReadClos
 		}
 	}
 
-	if offset != resumableOffset {
-		return storagedriver.InvalidOffsetError{Path: path, Offset: offset}
+	if offset < resumableOffset {
```

Comment:
@ahmetalpbalkan I think our conversation was along the lines of supporting overwrites of existing data at arbitrary offsets. Ultimately, operation is append-only, but we can have the writer dump data for offsets that are already present, reading back the missing block data if we have to assemble a new leading block.

The issue here is that we've gone to the more relaxed notion that an offset is only invalid if it is less than zero (we can pull this check up into the wrapper, perhaps). This is to avoid unexpectedly failed writes under eventually consistent backends.

There are two benefits to this functionality in the driver:

1. Clients that are "behind" what is actually written can continue writing from an old write position. This can happen if a client fails mid block and does get an updated token, but the block is written on the backend. Such a write should not fail.
2. Checkpointed checksum verification on uploads under an eventually consistent backend. This allows us to reduce the amount of data read back into the registry when verifying image data and moving it in the blob store.

I understand this behavior might be hard to implement but its benefits are clear for the client. We may want to build up some common code between the drivers to make this easier, but you can take a peak at the s3 driver to see how it was done.

That said, here's how we should proceed:

1. Implement a unit test that checks for writes before the end of a file and check that the azure driver fails the test and s3 and filesystem pass. Then, disable this test for now.

2. We can take one of two options:
  a. Allow for this driver to be append-only, knowing that it may result in odd failures in the future. Place in logging and metrics to prioritize future functionality in the azure driver. Leave the unit test disabled.
  b. Implement the full functionality and enable the unit test.